distrobuilder/chroot.go

```diff
@@ -4,80 +4,124 @@ import (
 	"fmt"
 	"os"
 	"path/filepath"
+	"regexp"
+	"strconv"
 	"syscall"
 
 	"github.com/lxc/distrobuilder/managers"
 	"github.com/lxc/distrobuilder/shared"
 )
 
-func prepareChroot(cacheDir string) ([]string, error) {
-	type Mount struct {
-		source string
-		target string
-		fstype string
-		flags  uintptr
-		data   string
+type chrootMount struct {
+	source  string
+	target  string
+	fstype  string
+	flags   uintptr
+	data    string
+	remove  bool
+	mounted bool
+	isDir   bool
+}
+
+func mountFilesystems(rootfs string, mounts []chrootMount) bool {
+	ok := true
+
+	for i, mount := range mounts {
+		if mount.isDir {
+			os.MkdirAll(filepath.Join(rootfs, mount.target), 0755)
+		} else {
+			os.Create(filepath.Join(rootfs, mount.target))
+		}
+		err := syscall.Mount(mount.source, filepath.Join(rootfs, mount.target),
+			mount.fstype, mount.flags, mount.data)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Failed to mount '%s': %s\n", mount.source, err)
+			ok = false
+			break
+		}
+		mounts[i].mounted = true
 	}
 
-	var (
-		err      error
-		unmounts []string
-	)
+	return ok
+}
 
-	mounts := []Mount{
-		{filepath.Join(cacheDir, "rootfs"), "", "tmpfs", syscall.MS_BIND, ""},
-		{"proc", "proc", "proc", 0, ""},
-		{"sys", "sys", "sysfs", 0, ""},
-		{"udev", "dev", "devtmpfs", 0, ""},
-		{"shm", "/dev/shm", "tmpfs", 0, ""},
-		{"/dev/pts", "/dev/pts", "tmpfs", syscall.MS_BIND, ""},
-		{"run", "/run", "tmpfs", 0, ""},
-		{"tmp", "/tmp", "tmpfs", 0, ""},
+func unmountFilesystems(rootfs string, mounts []chrootMount) {
```

Comment:
I think you could just do a lazy umount of the root of the mount path (rootfs) and that should clean everything up for you (if it doesn't, the mount namspace will take care of that).